CacheManager.StackExchange.Redis.SelfCleaning.Tests/SelfCleaningRedisIntegrationTests.cs

```diff
@@ -0,0 +1,247 @@
+﻿using System;
+using System.Collections.Concurrent;
+using System.Collections.Generic;
+using System.Linq;
+using System.Net;
+using System.Threading.Tasks;
+using CacheManager.Core;
+using CacheManager.Core.Internal;
+using CacheManager.StackExchange.Redis.SelfCleaning.Core;
+using CacheManager.StackExchange.Redis.SelfCleaning.Timers;
+using Moq;
+using NUnit.Framework;
+using StackExchange.Redis;
+
+namespace CacheManager.StackExchange.Redis.SelfCleaning.Tests
+{
+    [TestFixture]
+    public class SelfCleaningRedisIntegrationTests
+    {
+        private const double TIME_TO_LIVE_MILLISECONDS_DIFFERENCE_THRESHOLD = 100;
+
+        private Mock<IServer> _serverMock;
+        private Mock<IDatabase> _databaseMock;
+        private Mock<IConnectionMultiplexer> _connectionMock;
+
+        private IDictionary<RedisKey, (RedisValue[] Values, DateTime InsertionTime)> _fauxDatabase;
+
+        private TimeSpan _timeToLive;
+        private ICacheManager<DummyModel> _cache;
+
+        private ICollection<(CacheItemRemovedEventArgs Args, DateTime Time)> _onRemoveByHandleInvocations;
+
+        [OneTimeSetUp]
+        public void Setup()
+        {
+            _serverMock = new Mock<IServer>();
+            _databaseMock = new Mock<IDatabase>();
+            _connectionMock = new Mock<IConnectionMultiplexer>();
+
+            _fauxDatabase = new ConcurrentDictionary<RedisKey, (RedisValue[], DateTime)>();
+
+            SetupServerMock();
+            SetupDatabaseMock();
+            SetupConnectionMock();
+            SetupCache();
+        }
+
+        #region Setup Methods
+
+        private void SetupServerMock()
+        {
+            _serverMock.SetupGet(server => server.IsConnected).Returns(true);
+            _serverMock.SetupGet(server => server.Features).Returns(new RedisFeatures(Version.Parse("3.0.504")));
```

Comment:
Why specific 3.0.504? Does it affect something?

Reply:
The only thing this is relevant for is the Scripting feature. It is available starting at version 2.5.7. `RedisCacheHandle` does thing differently if Scripting isn't available. I've checked and it isn't too complicated to add setups for that case, but I wonder about the following things:
1. Do we need tests for both cases, or do we keep just one and necessarily use a matching version?
2. If we do have setups for both cases, how do we indicate which case we want to test? Add a `TestCase` for each test? Change it manually somehow?

Reply:
I dont think we need tests to versions we dont want to support, but we do need to explicitly say which versions we support.. maybe ReadME? Im not so sure, because I almost never really worked on infrastructure with opensources...

Reply:
I think the effort isn't that big, so I added setups for older versions (in which `RedisCacheHandle` uses `HashSet`\`HashGet` instead of `ScriptEvaluate`). You can change alter between versions with and without scripting by changing the `_useScripting` variable in the main setup. See how it looks

Reply:
But its a bit weird that there is "optional" boolean parameter in tests which should being changed by... us? The developers? In which case?

I think that we must figure out which option is more correct:
1. running two sets of tests which will have the same logic but different setup (for example, we can use BaseIntegrationTests with tests and just specific setup will be change)
2. We understand that we are not testing the version of redis, but our code. Therefore, we dont mind if the version is 3.0.504, 2.0.0 or 136.27.235, as long as we setup it as we should.

Personally I think that option 2 is better, thats outside the responsibility of our tests. Tell me if you think different

Reply:
Okay then. I take it I should keep both setups of ScriptEvaluate and HashSet/HashGet?

Reply:
If you are using version 3.0.504 so no, there is no reason for that. If you really want, you can detail more about it in the comment of the ScriptEvaluate, like "(if you want to use versions before 3.0.504. you'll need to setup HashSet and HashGet methods)"